workers/unity/Packages/com.improbable.gdk.core/Utility/PlayerLoopUtils.cs

```diff
@@ -0,0 +1,238 @@
+            var uniqueGroup = new HashSet<Type>(systems.Select(s => s.GetType()));
+
+            // Create simulation system for the default group
+            var simulationSystemGroup = world.GetOrCreateSystem<SimulationSystemGroup>();
+
+            // Add systems to their groups, based on the [UpdateInGroup] attribute.
+            for (var i = 0; i < systems.Count; i++)
+            {
+                var system = systems[i];
+                var type = system.GetType();
+
+                // Skip the root-level systems
+                if (type == typeof(InitializationSystemGroup) ||
+                    type == typeof(SimulationSystemGroup) ||
+                    type == typeof(PresentationSystemGroup) ||
+                    type.GetCustomAttribute<UpdateInSubSystemAttribute>() != null)
+                {
+                    uniqueGroup.Add(type);
+                    continue;
+                }
+
+                // Add to default group is none is defined
+                var groups = type.GetCustomAttributes(typeof(UpdateInGroupAttribute), true);
+                if (groups.Length == 0)
+                {
+                    simulationSystemGroup.AddSystemToUpdateList(system);
+                }
+
+                foreach (var group in groups)
+                {
+                    if (!(group is UpdateInGroupAttribute groupAttribute))
+                    {
+                        continue;
+                    }
+
+                    if (!typeof(ComponentSystemGroup).IsAssignableFrom(groupAttribute.GroupType))
+                    {
+                        Debug.LogError(
+                            $"Invalid [UpdateInGroup] attribute for {type}: {groupAttribute.GroupType} must be derived from ComponentSystemGroup.");
+                        continue;
+                    }
+
+                    var groupMgr = world.GetOrCreateSystem(groupAttribute.GroupType);
+                    if (groupMgr is ComponentSystemGroup groupSys)
+                    {
+                        groupSys.AddSystemToUpdateList(world.GetOrCreateSystem(type));
+                        if (!uniqueGroup.Contains(groupAttribute.GroupType))
+                        {
+                            uniqueGroup.Add(groupAttribute.GroupType);
+                            systems.Add(groupMgr);
+                        }
+                    }
+                }
+            }
+
+            // Sort all root groups, sorts depth first
+            foreach (var system in systems)
+            {
+                var type = system.GetType();
+                if (type == typeof(InitializationSystemGroup) ||
+                    type == typeof(SimulationSystemGroup) ||
+                    type == typeof(PresentationSystemGroup) ||
+                    type.GetCustomAttribute<UpdateInSubSystemAttribute>() != null)
+                {
+                    var groupSystem = system as ComponentSystemGroup;
+                    groupSystem.SortSystemUpdateList();
```

Comment:
This is wacky. Can we leave a comment explaining what this is doing?

Reply:
It does? `// Sort all root groups, sorts depth first`

Reply:
I guess I'm looking more for the why?

Reply:
This takes all the systems that were added to this group, and sorts them according to their [Before/After] group attributes.